stablehlo/reference/ProcessGrid.cpp

```diff
@@ -228,5 +278,19 @@ std::shared_ptr<RendezvousResult const> ProcessGrid::rendezvous(
   return result;
 }
 
+void ProcessGrid::send(ArrayRef<Tensor> inputs, ChannelId channelId,
+                       ProcessId processId) {
+  std::unique_lock<std::mutex> lock(sendRecvChannels_[channelId].mutex);
+  if (!sendRecvConditions_[channelId].wait_for(
+          lock, std::chrono::seconds(3), [&] {
+            return sendRecvReady_.find(channelId) != sendRecvReady_.end();
```

Comment:
Can we add a contains method? This code currently:
- Gets mutex, calls begin, releases mutex
- Gets mutex, calls end, releases mutex

There is no guarantee (very unlikely though) that the set is unmodified in the meantime? Having a contains function should make this a single mutex event

Reply:
Also curious if we should check if receive is ready prior to waiting? In case the `recv` thread reached this point first?

Reply:
Note that the `wait` function will release the lock when the condition is not satisfied. If send happens first, it will lock, and wait (release the lock). Then recv will then insert to `sendRecvReady_` and notify, at which point send's wait condition is satisfied. send will append to result and notify, at which point recv similarly will reacquire the lock and see that the condition is satisfied and return.